Blog gsoc26 xpand2 (#572)
unikraft/docs: content/blog/2026-07-14-unikraft-gsoc-xpand-2.mdx

---
title: "GSoC'26: Expanding the Unikraft Software Support Ecosystem"
description: |
 This is IInd in series to my blogs on expanding the app ecosystem on Unikraft.
publishedDate: 2026-07-14
image: /images/unikraft-gsoc24.png
authors:
- Tushar Verma
tags:
- gsoc
- gsoc26
- virtualization
- apps
---

## Project Overview

> This is the IInd in my series blog on expanding the app ecosystem on Unikraft.

Reviewer: ```suggestion
This is the second blog post related to my GSoC project on expanding the app ecosystem on Unikraft.
```

> In Part I, I covered dnsmasq ,lighttpd(bincompat-port) ,nginx update and other things.

Reviewer: Please check comma spacing everywhere in this post: no space before comma, one space after comma

Reviewer: Don't use "other things", be specific

## Progress

### lighttpd

[lighttpd](https://www.lighttpd.net/), a lightweight web server, now has a native port!
The earlier route was binary-compatibility mode — package the binary and its shared libraries into a rootfs and ran under the ELF loader.

> The deeper route is a native source port which is often more performant, `lib-lighttpd`: lighttpd `1.4.84` compiled directly into the unikernel against musl and lwIP.

Reviewer: `lib-lighttpd`: lighttpd `1.4.84` -> it is not clear for the reader, who compiled, the lib or the app?

**Escaping Meson**
lighttpd builds with meson/cmake/automake , all of which generate files before compiling: a `config.h` of platform feature probes, and a configuration-file parser produced by [lemon](https://sqlite.org/lemon.html) (SQLite's parser generator) from a grammar file.
Unikraft's build is plain make and fetches a tarball, so none of that generation happens.
The port ships those artifacts pre-generated in the library repo instead: a hand-written `config.h` describing what musl + lwIP actually provide (epoll via `posix-poll`, IPv6, `getrandom`; no `fork`, no `sendfile`), and `configparser.c` taken straight from midway through meson's build pipeline.

**TLS as a Kconfig option.**
`mod_openssl` compiles against Unikraft's `lib-openssl` behind a single `LIBLIGHTTPD_OPENSSL` option, so HTTPS support — and the whole OpenSSL dependency — is one menuconfig toggle.
`lib-openssl` still ships OpenSSL `1.1.1c`; moving it to a current `3.x` LTS is a heavy port of its own (provider model, ~900 source files), left for future work.

> One entry-point subtlety worth recording: lighttpd's `server.c` wraps its `main` in `#ifndef main / #define server_main main`, so the usual `-Dmain=app_main` rename trick silently leaves the symbol as `server_main` — the port's `main()` wrapper calls that directly.

Reviewer: ```suggestion
One entry-point subtlety worth recording: lighttpd's `server.c` wraps its `main()` in `#ifndef main / #define server_main main`, so the usual `-Dmain=app_main` rename trick silently leaves the symbol as `server_main` — the port's `main()` wrapper calls that directly.
```

> One entry-point subtlety worth recording: lighttpd's `server.c` wraps its `main` in `#ifndef main / #define server_main main`, so the usual `-Dmain=app_main` rename trick silently leaves the symbol as `server_main` — the port's `main()` wrapper calls that directly.

Reviewer: Add a section title for this paragraph, as you did for the other 2 in this section

**Related PRs:** [lib-lighttpd](https://github.com/unikraft/lib-lighttpd/pull/1)

### Library housekeeping

Native ports keep tripping over the same class of outdated libs , so some fixes landed in the libraries themselves:

**`lib-openssl`: dead download URL.**
`openssl.org` retired its `/source/old/` archive, so build failed at the fetch step with a tarball that no longer existed.
The fetch URL now points at the GitHub release tag, which is permanent.

**`lib-pcre`: dead download URL.**

> In a simillar fashion to openssl , the archive moved from `pcre-org` to `sourceforge`.

Reviewer: Please only use self-contained sentences, don't rely on the title to infer the subject.
In this case, In a simillar fashion to openssl , the archive for `lib-pcre` moved from `pcre-org` to `sourceforge`.

**`lib-zlib`: hard-wired to the deprecated filesystem stack.**

> Its `Config.uk` hard-selected `vfscore` components, which breaks any image built on the new posix-vfs stack.

Reviewer: Same remark as in the previous parapgraph, don't rely on the title when writing the parapgraph; instead of `Its`, use "The `Config.uk` for `lib-zlib` hard-selected...`

The result is an NTP server appliance: boot it, it syncs to the pool, and every machine on the network can get accurate time from it.

It runs in binary-compatibility mode , but getting there surfaced some gaps.

**An empty `resolv.conf`, injected by the loader.**
With output visible, the syscall trace showed DNS queries going nowhere: `read(/etc/resolv.conf) = 0`.
The elfloader's host-filesystem helper (`APPELFLOADER_HFS_ETCRESOLVCONF`) is supposed to inject the host's resolver configuration, but under `kraft run` it creates an *empty* file — and with `HFS_REPLACEEXIST` it even overwrites the one shipped in the rootfs.
Disabling the helper and baking a static `resolv.conf` (public resolvers;  queried in parallel) fixed upstream resolution.

**No `recvmmsg` in Unikraft.**
Replies from the pool arrived, `select()` reported the socket readable, and chronyd still never read a packet.
Alpine's chrony is compiled with `HAVE_RECVMMSG` and has no runtime fallback, while Unikraft's syscall shim returns `-ENOSYS` — so chronyd busy-looped on a socket it could never consume.

> I verified two fixes: an `LD_PRELOAD` shim that rewrites `recvmmsg` into a `recvmsg` loop (keeps the stock distro binary), and building chrony from source with `HAVE_RECVMMSG` removed from `config.h`.

Reviewer: Which is the solution you sticked to in the final PR and why? This should be mentioned here

The port ships the source build , it also lets me drop features that are dead weight in a unikernel (`cmdmon`, refclocks, privilege dropping, seccomp) and strip the binary down to ~250 KB.

**A dual-stack compile bug in lib-lwip.**
Enabling IPv6 broke the build in Unikraft's netlink glue: `netif->address.type` — a field that doesn't exist; lwIP calls it `ip_addr`.

> The broken line only compiles when netlink and IPv6 are enabled together .

Reviewer: ```suggestion
The broken line only compiles when netlink and IPv6 are enabled together.
```

One-word fix, headed upstream. (IPv6 stayed off in the shipped images: with `AF_INET6` sockets available chrony preferred the unreachable v6 addresses of every upstream and never synced.)

**One decision worth explaining: two images.**
The port ships as `chronyd/4.8` and `chronyd/4.8-nts`.
NTS (Network Time Security) authenticates the NTP exchange over TLS, which is a real upgrade — but it drags in gnutls and its whole dependency closure plus a CA bundle, inflating the rootfs from **~2 MB to ~10 MB**.
Deployments syncing within trusted infrastructure shouldn't pay a 5x size cost for certificates they don't check, so the plain image stays minimal and the NTS image is there when authenticated time matters.

> Both were verified end-to-end: the plain image syncs to the pool and answers clients at stratum 2; the NTS image completes NTS-KE against Cloudflare and PTB (All configurable).

Reviewer: ```suggestion
Both were verified end-to-end: the plain image syncs to the pool and answers clients at stratum 2; the NTS image completes NTS-KE against Cloudflare and PTB (all configurable).
```

## Next Steps

- Upstream the lighttpd's catalog-core app.

> - Upstream a [Odin](https://odin-lang.org/) hello-world program to catalog-core.

Reviewer: ```suggestion
- Upstream an [Odin](https://odin-lang.org/) hello-world program to catalog-core.
```

- Upstream the fixes that fell out of the chrony port:`recvmmsg` in `lib/posix-socket`, the lib-lwip netlink dual-stack fix .
- Investigate [Lwip'sntp application module](https://www.nongnu.org/lwip/2_0_x/group__sntp.html).

> - Continue down the project list: Investigate Gitea, and a message broker.

Reviewer: ```suggestion
- Continue down the project list: Investigate Gitea and a message broker.
```

---
title: "GSoC'26: Expanding the Unikraft Software Support Ecosystem"
description: |

>  This is IInd in series to my blogs on expanding the app ecosystem on Unikraft.

Reviewer: ```suggestion
This is the second blog post related to my GSoC project on expanding the app ecosystem on Unikraft.
```
